Made numeric codec implementations more lenient

rozza opened the pull request:
The leniency covers the following BsonTypes: DOUBLE, INT32 and INT64.

Its relatively simple compared to the conversions available in c#. I'm not sure we want to expand this further as I don't see users mixing types with Decimals in the same way as they might with ints and longs.

JAVA-2529

rstam commented on bson/src/main/org/bson/codecs/AtomicIntegerCodec.java:
```diff
@@ -21,21 +21,47 @@
+     * Construct a new instance
+     *
+     * @param delta the maximum delta between {@code expected} and {@code actual} for which both numbers are still
+     * considered equal. Required when converting Double values to {@code AtomicInteger} values. Defaults to {@code 0.00000000000001d}.
```
Why would anyone ever want a delta other than 0.0? That just results in information loss.

rstam commented on bson/src/main/org/bson/codecs/NumberCodecHelper.java:
```diff
@@ -0,0 +1,82 @@
+                break;
+            case INT64:
+                number = reader.readInt64();
+                isADownConversion = clazz == Integer.class;
```
Conversion to double can lose information also. There are certain very large INT64 numbers that will lose precision when converted to double.

rstam commented on bson/src/main/org/bson/codecs/NumberCodecHelper.java:
```diff
@@ -0,0 +1,82 @@
+        return (T) converted;
+    }
+
+    private static <T extends Number> boolean hasLostPrecision(final double d1, final double d2, final double delta) {
```
I don't think this works for large INT64 values that don't fit in a double.

rstam commented on bson/src/main/org/bson/codecs/NumberCodecHelper.java:
```diff
@@ -0,0 +1,82 @@
+
+final class NumberCodecHelper {
+
+    static final double DEFAULT_DELTA = 0.00000000000001d;
```
Also, decimal fractions rarely convert well to binary-based doubles.

rozza commented:
Thanks @rstam - thats embarrassing. Looks like my lack of formal training and adapting a stackoverflow answer doesn't fit the need.

I'll update and message once ready.

jyemin commented:
@rozza ready for second pass?

CC @rstam

rozza commented:
Yes.  I've updated the code and simplified it. I now test that we don't lose fidelity by adding a round trip correctness test in the `NumberCodecHelper`.

jyemin commented:
@rstam please take another pass at this.

jyemin commented on bson/src/test/unit/org/bson/codecs/DoubleCodecTest.java:
```diff
@@ -0,0 +1,40 @@
+    public void shouldHandleAlternativeNumberValues() {
+        Document expected = new Document("a", 10.00);
+        roundTrip(new Document("a", 10), expected);
+        roundTrip(new Document("a", 10L), expected);
```
Add some tests at the limit:

        roundTrip(new Document("a", Long.MAX_VALUE), new Document("a", (double) Long.MAX_VALUE));
        roundTrip(new Document("a", Long.MIN_VALUE), new Document("a", (double) Long.MIN_VALUE));

jyemin commented on bson/src/test/unit/org/bson/codecs/DoubleCodecTest.java:
```diff
@@ -0,0 +1,40 @@
+        roundTrip(new Document("a", 10), expected);
+        roundTrip(new Document("a", 10L), expected);
+    }
+
```
Add some tests for lossy longs:

    @Test(expected = BsonInvalidOperationException.class)
    public void shouldThrowWhenHandlingLossyLongValues() {
        roundTrip(new Document("a", Long.MAX_VALUE - 1));
        roundTrip(new Document("a", Long.MIN_VALUE + 1));
    }

jyemin commented on bson/src/main/org/bson/codecs/NumberCodecHelper.java:
```diff
@@ -0,0 +1,97 @@
+final class NumberCodecHelper {
+
+    @SuppressWarnings("unchecked")
+    static <T extends Number> T decodeNumber(final BsonReader reader, final Class<T> clazz) {
```
I think it's worth breaking up this method into three specialized ones, one for each of the three supported numeric types.  This will avoid warning suppression, unnecessary primitive boxing/unboxing, redundant casts, and checks that are unnecessary if you know the expected type.  As an example for double, I'd prefer something more like this:

    static double decodeDouble(final BsonReader reader) {
        double doubleValue;
        BsonType bsonType = reader.getCurrentBsonType();
        switch (bsonType) {
            case INT32:
                doubleValue = reader.readInt32();
                break;
            case INT64:
                long longValue = reader.readInt64();
                doubleValue = longValue;
                if (longValue != (long) doubleValue) {
                    throw invalidConversion(Double.class, longValue);
                }
                break;
            case DOUBLE:
                doubleValue = reader.readDouble();
                break;
            default:
                throw new BsonInvalidOperationException(format("Invalid numeric type, found: %s", bsonType));
        }
        return doubleValue;
    }

rstam replied:
I agree.

jyemin replied:
Now that the specializations are in, this method is unused and can be deleted.

jyemin commented on bson/src/main/org/bson/codecs/AtomicIntegerCodec.java:
```diff
@@ -21,21 +21,30 @@
 
 public class AtomicIntegerCodec implements Codec<AtomicInteger> {
+
+    /**
```
Nit... but I think these are just left over from the first patch.

rozza commented:
I added a Pojo test as well. I think this covers all items, but before merging is there any benefit to adding the coercion of strings to numerics?

jyemin commented on bson/src/main/org/bson/codecs/NumberCodecHelper.java:
```diff
@@ -163,8 +97,8 @@ static double decodeDouble(final BsonReader reader) {
 
-    private static  <T extends Number> BsonInvalidOperationException invalidConversion(final Class<T> clazz, final Number number) {
-        return new BsonInvalidOperationException(format("Could not convert number (%s) to %s without losing precision", number, clazz));
+    private static  <T extends Number> BsonInvalidOperationException invalidConversion(final Class<T> clazz, final Object value) {
```
This is still always a Number, isn't it?

rozza commented:
Rebased and merged
